### PR TITLE
Align federated helmrepositories.source.toolkit.fluxcd.io observedGeneration semantics with its native

### LINKED ISSUE #4870: [Umbrella] Align generation and observedGeneration fields by reflectstatus and aggregatestatus operation
**What would you like to be added**:
There are some third-party resources and native resources that need to align `.metadata.generation`  and `.status.observedGeneration ` fields. Please ref to #4866 .

**Why is this needed**:

Continue the work of #4867  and break down the tasks to be done.
- Native resource
  - [x] GroupVersion: apps/v1
    Kind: 
      - [x] Deployment @veophi #4867 
      - [x] DaemonSet @whitewindmills #5165 
      - [x] StatefulSet @liangyuanpeng #5204
- Thirdparty resource
  - [x] GroupVersion: apps.kruise.io/v1alpha1
    Kind: 
      - [x] CloneSet @veophi #5057
      - [x] DaemonSet @whitewindmills #5167 
  - [x] GroupVersion: apps.kruise.io/v1beta1
    Kind: 
      - [x] StatefulSet @zhy76 #5094
  - [x] GroupVersion: helm.toolkit.fluxcd.io/v2beta1
    Kind: 
      - [x] HelmRelease @yike21 #5311 
  - [x] GroupVersion: kustomize.toolkit.fluxcd.io/v1
    Kind: 
      - [x] Kustomization @yike21 #5084 
  - [x] GroupVersion: source.toolkit.fluxcd.io/v1
    Kind: 
      - [x] GitRepository @yike21 #5086 
  - [x] GroupVersion: source.toolkit.fluxcd.io/v1beta2
    Kind: 
     - [x] Bucket @yike21  #5193 
     - [x] HelmChart @yike21 #5194
     - [x] HelmRepository @yike21 #5196
     - [x] OCIRepository @yike21 #5197

We need to examine the meaning of `.metadata.generation` and `.status.observedGeneration` fields for each resource above, and then modify the `ReflectStatus` and `AggregateStatus` operations to align them. The modifications are likely to be similar, so please wait for `CloneSet` and `Deployment` modifications to be completed.

/help

### PR DESCRIPTION
**Pull request by yike21**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #4870 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
`karmada-controller-manager`: Mark `.status.observedGeneration` of helmrepositories.source.toolkit.fluxcd.io with `.metadata.generation` only when all members' statuses are algined with its resource template generation.
```

### CODE REVIEW

File: pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/source.toolkit.fluxcd.io/v1beta2/HelmRepository/customizations.yaml [hunk 1]
```diff
@@ -22,22 +22,42 @@ spec:
...
+          -- Count all members that their status is updated to the latest generation
+          local observedResourceTemplateGenerationCount = 0
+          
           local conditionsIndex = 1
           for i = 1, #statusItems do
             if statusItems[i].status ~= nil and statusItems[i].status.artifact ~= nil then
-              desiredObj.status.artifact = statusItems[i].status.artifact
-            end
-            if statusItems[i].status ~= nil and statusItems[i].status.lastHandledReconcileAt ~= nil and statusItems[i].status.lastHandledReconcileAt ~= '' then
-              desiredObj.status.lastHandledReconcileAt = statusItems[i].status.lastHandledReconcileAt
```
Comment (XiShanYongYe-Chang) [inline pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/source.toolkit.fluxcd.io/v1beta2/HelmRepository/customizations.yaml:37]: These two lines are the extra settings from before, right?

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thanks
/lgtm
/assign @chaunceyjiang

File: pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/source.toolkit.fluxcd.io/v1beta2/HelmRepository/customizations.yaml (hunk 1 above)
Comment (yike21) [reply pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/source.toolkit.fluxcd.io/v1beta2/HelmRepository/customizations.yaml:37]: Yes, I find that this field is not in the current `helmrepositories.source.toolkit.fluxcd.io` API. 
It appears to be a past bug that wasn't  triggered yet.

**Review by RainbowMango — APPROVED** [review]
/approve

**Pull request merged**